File: tests/drivers/go/v5/parallel_edge_import.go
package main

import (
	"log"
    "fmt"
	"sync"
    "github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

var nodeCount int = 100

func handle_if_error(err error) {
	if err != nil {
	  log.Fatal("Error occured: %s", err)
	}
}

func main() {
    dbUri := "bolt://localhost:7687"
    driver, err := neo4j.NewDriver(dbUri, neo4j.BasicAuth("", "", ""))
    if err != nil {
        log.Fatal("An error occurred opening conn: %s", err)
    }
    defer driver.Close()

	session := driver.NewSession(neo4j.SessionConfig{})
	defer session.Close()

	_, err = session.WriteTransaction(clearDatabase)
	handle_if_error(err)
	fmt.Println("Database cleared.")

	_, err = session.WriteTransaction(createRoot)
	handle_if_error(err)
	fmt.Println("Record created.")

	parallelEdgeImport(driver)

	_, err = session.WriteTransaction(testNodeCount)
	handle_if_error(err)
	fmt.Println("All ok!")
}

func clearDatabase(tx neo4j.Transaction) (interface{}, error) {
	result, err := tx.Run(
        "MATCH (n) DETACH DELETE n;",
        map[string]interface{}{})
	handle_if_error(err)
	fmt.Printf("Database cleared!\n")
	return result.Consume()
}

func createRoot(tx neo4j.Transaction) (interface{}, error) {
    result, err := tx.Run(
        `MERGE (root:Root) RETURN root;`,
        map[string]interface{}{})
	handle_if_error(err)
	fmt.Printf("Root created!\n")
    return result.Consume()
}

func worker(driver neo4j.Driver, wg *sync.WaitGroup) {
    defer wg.Done() // Decrement the WaitGroup counter when the goroutine is done

	session := driver.NewSession(neo4j.SessionConfig{})
	defer session.Close()

	_, err := session.WriteTransaction(edgeImport)
	handle_if_error(err)
    fmt.Printf("Transaction got through!\n")
}

func parallelEdgeImport(driver neo4j.Driver)  {
    var wg sync.WaitGroup

    for i := 1; i <= nodeCount - 1; i++ {
        wg.Add(1)
        go worker(driver, &wg)
    }

    wg.Wait()

    fmt.Println("All workers have finished.")
}

func edgeImport(tx neo4j.Transaction) (interface{}, error) {
    result, err := tx.Run(
        `MATCH (root:Root) CREATE (n:Node) CREATE (n)-[:TROUBLING_EDGE]->(root);`,
        map[string]interface{}{})
	handle_if_error(err)
    return result.Consume()
}

func testNodeCount(tx neo4j.Transaction) (interface{}, error) {
	result, err := tx.Run(
        "MATCH (n) RETURN COUNT(n) AS cnt;",
        map[string]interface{}{})
	handle_if_error(err)

	if !result.Next() {
		log.Fatal("Missing result.")
	}

	count, found := result.Record().Get("cnt")
	if !found {
	  return nil, fmt.Errorf("Wrong result returned.")
	}

    var countInt int
    switch v := count.(type) {
    case int:
        countInt = v
    case int64:
        countInt = int(v)
    case float64:
        countInt = int(v)
    default:
        return nil, fmt.Errorf("Unexpected data type for count: %T", count)
    }

	if countInt != nodeCount {
		log.Fatal("Count does not match! (", count, ")")
	}

    return result.Consume()
}

File: tests/drivers/java/v5_8/src/main/java/memgraph/ParallelEdgeImport.java
package memgraph;

import static org.neo4j.driver.Values.parameters;

import java.util.*;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.CompletableFuture;
import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.Config;
import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Result;
import org.neo4j.driver.Session;
import org.neo4j.driver.Transaction;
import org.neo4j.driver.TransactionWork;
import org.neo4j.driver.async.AsyncSession;
import org.neo4j.driver.async.AsyncTransaction;
import org.neo4j.driver.async.AsyncTransactionWork;
import org.neo4j.driver.async.AsyncSession;
import org.neo4j.driver.exceptions.ClientException;
import org.neo4j.driver.exceptions.TransientException;

public class ParallelEdgeImport {
  public static int dropDatabase(Driver driver) {
    try (Session session = driver.session()) {
      session.writeTransaction(new TransactionWork<Integer>() {
        @Override
        public Integer execute(Transaction tx) {
          tx.run("MATCH (n) DETACH DELETE n;");
          return 0;
        }
      });
    } catch (ClientException e) {
      System.out.println(e);
    }

    return 0;
  }

  public static int createRoot(Driver driver) {
    try (var session = driver.session()) {
      session.writeTransaction(new TransactionWork<Integer>() {
        @Override
        public Integer execute(Transaction tx) {
          tx.run("MERGE (n:Root) RETURN ID(n);");
          return 0;
        }
      });
    } catch (ClientException e) {
      System.out.println(e);
    }

    return 0;
  }

  public static void parallelEdgeImport(Driver d, int nodeCount) {
    try (var session = driver.session()) {
      session.writeTransaction(new TransactionWork<Integer>() {
        @Override
        public Integer execute(Transaction tx) {
          tx.run("MATCH (root:Root) CREATE (n:Node) CREATE (n)-[:TROUBLING_EDGE]->(root);");
          System.out.println("Transaction got through!");
          return 0;
        }
      });
    } catch (ClientException e) {
      System.out.println(e);
    }
  }

  private static Config config = Config.builder()
                        .withoutEncryption()
                        .withMaxTransactionRetryTime(15, TimeUnit.SECONDS)
                        .build();

  private static Driver driver = GraphDatabase.driver("bolt://localhost:7687", AuthTokens.none(), config);
  private static int nodeCount = 100;

  static class MyRunnable implements Runnable {
    @Override
    public void run() {
        parallelEdgeImport(driver, nodeCount);
    }
  }

  public static void main(String[] args) {
    dropDatabase(driver);
    System.out.println("Database cleared!");

    createRoot(driver);
    System.out.println("Root created!");

    List<Thread> threadList = new ArrayList<>();
    for (int i = 0; i < nodeCount - 1; i++) {
        Runnable myRunnable = new MyRunnable();
        Thread thread = new Thread(myRunnable);
        threadList.add(thread);
        thread.start();
    }

    // Wait for all threads to finish
    for (Thread thread : threadList) {
      try {
          thread.join();
      } catch (InterruptedException e) {
          e.printStackTrace();
      }
    }

    try (var session = driver.session()) {
      var actualNodeCount = session.run("MATCH (n) RETURN COUNT(n) AS cnt;").list().get(0).get("cnt").asInt();
      if (actualNodeCount != nodeCount) {
        System.out.println("Node count doesn't match (" + actualNodeCount + ")");
        System.exit(1);
      }
    }

    System.out.println("All ok!");
    driver.close();
  }
}
